keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Constraint.kt

```diff
@@ -1,14 +1,36 @@
 package com.netflix.spinnaker.keel.api
 
-import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.fasterxml.jackson.annotation.JsonSubTypes
+import com.fasterxml.jackson.annotation.JsonSubTypes.Type
+import com.fasterxml.jackson.annotation.JsonTypeInfo
 
-@JsonDeserialize(`as` = DependsOnConstraint::class)
-sealed class Constraint
+@JsonTypeInfo(
+    include = JsonTypeInfo.As.EXISTING_PROPERTY,
+  use = JsonTypeInfo.Id.NAME,
+  property = "type")
+@JsonSubTypes(
+  Type(value = DependsOnConstraint::class, name = "depends-on"),
+  Type(value = TimeWindowConstraint::class, name = "allowed-times")
+)
+sealed class Constraint(val type: String)
 
 /**
  * A constraint that requires that an artifact has been successfully promoted to a previous
  * environment first.
  */
 data class DependsOnConstraint(
   val environment: String
-) : Constraint()
+) : Constraint("depends-on")
+
+/**
+ * A constraint that requires the current time to fall within an allowed window
+ */
+data class TimeWindowConstraint(
+  val windows: List<TimeWindow>,
+  val tz: String? = null
+) : Constraint("allowed-times")
+
+data class TimeWindow(
+  val days: String? = null,
+  val hours: String? = null
+)
```

Comment:
I think defining these here as strings means we get no validation when you submit a delivery-config. Do you think we could do some sort of validation like we do [here](https://github.com/spinnaker/keel/blob/master/keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Resource.kt#L30), where that validation is pulled from the `AllowedTimesConstraintEvaluator`? 

I'd like to know when I submit that I've spelled something wrong vs an error later :)